x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1137,6 +1141,49 @@ protected Expression rule(Expression e) {
         }
     }
 
+    public static final class SimplifyArithmeticsInBinaryComparisons extends OptimizerExpressionRule {
+
+        public SimplifyArithmeticsInBinaryComparisons() {
+            super(TransformDirection.UP);
+        }
+
+        @Override
+        protected Expression rule(Expression e) {
+            if (e instanceof BinaryComparison) {
+                return reduce((BinaryComparison) e);
+            }
+            return e;
+        }
+
+        private Expression reduce(BinaryComparison bc) {
+            // optimize only once the expression looks like: AnyNode [ArithmOp] AnyNode [BiComp] Literal
+            if (bc.right() instanceof Literal == false) {
+                return bc;
+            }
+            Literal bcRight = (Literal) bc.right();
+
+            if (bc.left() instanceof ArithmeticOperation) {
+                ArithmeticOperation opLeft = (ArithmeticOperation) bc.left();
+                if (opLeft instanceof Mod) {
+                    return bc; // can't optimise Mods
+                }
+
+                if (opLeft.left() instanceof Literal || opLeft.right() instanceof Literal) { // 5 [op] a >= 4 || a [op] 5 >= 4
+                    // force double division
+                    if (opLeft.symbol().equals("*")) { // use symbol comp for SQL's Mul
+                        bcRight = new Literal(bcRight.source(), ((Number) bcRight.value()).doubleValue(), DOUBLE);
+                    }
+                    return bc.replaceChildren(List.of(opLeft.left(), opLeft.inverse(bcRight.source(), bcRight, opLeft.right())));
```

Comment:
Precision loss might cause some issues here. `a + b == c` is mathematically the same as `a == c - b`, except when you take into account the potential precision loss.

For example:
```
a = 1.0E-15
b = 1.0E15
c = 1.0E15

a + b == c
1.0E-15 + 1.0E15 == 1.0E15
1.0E15 == 1.0E15
true

a == c - b
1.0E-15 == 1.0E15 - 1.0E15
1.0E-15 == 0.0
false
```

So even though the query before and after the optimization might be mathematically equivalent, it still won't return the same result after the optimization is applied. This might have unintended consequences after a version upgrade. I think we should discuss this before moving ahead with the PR.

Reply:
Thanks, good point. Addressed, along with the similar DIV/MUL concern.